### PR TITLE
[refactor] Pass element_shape and layout to C++ Ndarray

### PR DESCRIPTION
**Pull request by ailzhang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #5073
* __->__ #5066
* #5065

Note we still flatten element_shape in the C++ Ndarray, which is blocked by the accessors and will be fixed
in the following PRs.

### CODE REVIEW

**Comment by turbo0628** [comment]
It saves lots of guess works in the codegen! Looking forward to the changes!

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 1]
```diff
@@ -486,13 +486,14 @@ class KernelGen : public IRVisitor {
...
-    const auto layout = stmt->element_dim <= 0 ? layout_AOS : layout_SOA;
+
+    const auto layout = stmt->element_dim <= 0 ? ExternalArrayLayout::kAOS
+                                               : ExternalArrayLayout::kSOA;
 
     if (element_shape.size() > 0) {
       int elem_beg = 0;
       int elem_end = 0;
-      if (layout == layout_SOA) {
+      if (layout == ExternalArrayLayout::kSOA) {
```
Comment (k-ye) [inline taichi/backends/opengl/codegen_opengl.cpp:496]: IIUC, once element_shape is no longer passed into extra_args, we can  remove this part one...? (Metal now has the same stuff due to the Ndarray support) cc @turbo0628

**Review by k-ye — APPROVED** [review]
LGTM!

File: taichi/backends/opengl/codegen_opengl.cpp (hunk 1 above)
Comment (ailzhang) [reply taichi/backends/opengl/codegen_opengl.cpp:496]: Yup we briefly discussed this in the afternoon and @turbo0628 will get rid of these hacks later (still need some more cleanup before that

**Pull request merged**